### PR TITLE
Prevent allowhtml attribute

### PR DESCRIPTION
**Pull request by ViktorijaAdo**
Added analysis for allowHtml attribute for Properties and also tried to use SyntaxNodeHelper to avoid repeated code.

### CODE REVIEW

File: SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs [hunk 1]
```diff
@@ -7,6 +7,8 @@
 using VB = Microsoft.CodeAnalysis.VisualBasic;
 using VBSyntax = Microsoft.CodeAnalysis.VisualBasic.Syntax;
 
+using SecurityCodeScan.Analyzers.Utils;
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs]: Could you please move the using up like:
```cs
using SecurityCodeScan.Analyzers.Locale;
using SecurityCodeScan.Analyzers.Utils;
using CSharp = Microsoft.CodeAnalysis.CSharp;
```

File: SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs [hunk 2]
```diff
@@ -18,42 +20,77 @@ public class RequestValidationAnalyzer : DiagnosticAnalyzer
         public override void Initialize(AnalysisContext context)
         {
             // Separated the parsers on this one as they use too many language dependent syntax types. 
-            // TODO: Review to see if this can be simplified.
+            // TODO: Review to see if this can be simplified.\
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs]: `\` is a typo

File: SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs [hunk 3]
```diff
@@ -18,42 +20,77 @@ public class RequestValidationAnalyzer : DiagnosticAnalyzer
...
+            context.RegisterSyntaxNodeAction(VisitPropertiesCSharp,         CSharp.SyntaxKind.PropertyDeclaration);
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs]: There is even more elegant way not introduce additional classes or functions - lambda, like:
`context.RegisterSyntaxNodeAction(ctx => VisitProperties(ctx, CSharpSyntaxNodeHelper.Default),   CSharp.SyntaxKind.PropertyDeclaration);`

File: SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs [hunk 4]
```diff
@@ -18,42 +20,77 @@ public class RequestValidationAnalyzer : DiagnosticAnalyzer
...
+        {
+            var attributes = nodeHelper.GetPropertyAttributeNodes(ctx.Node);
+
+            SyntaxNode allowHtmlAttribute = null;
+            foreach (var attribute in attributes)
+            {
+                if (nodeHelper.GetAttributeNameNode(attribute).ToString().Contains("AllowHtml"))
+                {
+                    allowHtmlAttribute = attribute;
+                    break;
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs]: While it is corner case and unlikely to happen, there is flaw that it expects only one attribute that contains AllowHtml, so it breaks out the loop on the first occurrence, that can be a substring from a wrong attribute.
Could you also add a test for it?

File: SecurityCodeScan/Analyzers/Utils/CSharpSyntaxNodeHelper.cs [hunk 5]
```diff
@@ -157,6 +157,22 @@ public override SyntaxNode GetDefaultValueForAnOptionalParameter(SyntaxNode decl
...
+            if (!(node is AttributeSyntax attribute))
+                return null;
+
+            return attribute.Name;
+        }
+
+        public override SyntaxToken GetPropertyIdentifierNode(SyntaxNode node)
+        {
+            if (!(node is PropertyDeclarationSyntax property))
+                return new SyntaxToken();
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Utils/CSharpSyntaxNodeHelper.cs]: It is better to make return type nullable `SyntaxToken` and return null instead.

File: SecurityCodeScan/Analyzers/Utils/VBSyntaxNodeHelper.cs [hunk 6]
```diff
@@ -155,6 +155,22 @@ public override SyntaxNode GetCallTargetNode(SyntaxNode node)
...
+            if (!(node is AttributeSyntax attribute))
+                return null;
+
+            return attribute.Name;
+        }
+
+        public override SyntaxToken GetPropertyIdentifierNode(SyntaxNode node)
+        {
+            if (!(node is PropertyStatementSyntax property))
+                return new SyntaxToken();
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Utils/VBSyntaxNodeHelper.cs]: It is better to make return type nullable `SyntaxToken` and return null instead.

File: SecurityCodeScan.Test/Tests/RequestValidationAnalyzerTest.cs [hunk 7]
```diff
@@ -390,6 +390,59 @@ End Namespace
...
+End Namespace
+";
+
+            var expected = new DiagnosticResult
+            {
+                Id = "SCS0017",
+                Severity = DiagnosticSeverity.Warning
+            };
+
+            await VerifyCSharpDiagnostic(cSharpTest, expected.WithLocation("Test0.cs", 14, 23)).ConfigureAwait(false);
```
Comment (JarLob) [inline SecurityCodeScan.Test/Tests/RequestValidationAnalyzerTest.cs]: I missed that during the first review, but it should report an attribute not the property name. The test passes even without your loop/break fix, but it should fail until the analyzer is fixed.

File: SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs [hunk 8]
```diff
@@ -33,28 +33,22 @@ private void VisitProperties(SyntaxNodeAnalysisContext ctx, SyntaxNodeHelper nod
         {
             var attributes = nodeHelper.GetPropertyAttributeNodes(ctx.Node);
 
-            SyntaxNode allowHtmlAttribute = null;
             foreach (var attribute in attributes)
             {
                 if (nodeHelper.GetAttributeNameNode(attribute).ToString().Contains("AllowHtml"))
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs]: Please invert the `if` like:
```cs
if(!contain)
    continue;
```
It reduces nesting and makes the code more readable.

File: SecurityCodeScan.Test/Tests/RequestValidationAnalyzerTest.cs (hunk 7 above)
Comment (ViktorijaAdo) [reply SecurityCodeScan.Test/Tests/RequestValidationAnalyzerTest.cs]: Do you mean that this test DetectAllowHtmlAttributeAfterAtrributeContainingAllowHtmlInName passed before my last commit? Cause it's failing for me 
image

Comment (JarLob) [reply SecurityCodeScan.Test/Tests/RequestValidationAnalyzerTest.cs]: Hmm, you are right, it fails if I revert a changeset. I tried to revert the loop manually and it didn't fail before.
Still I think it is better to highlight the attribute.

**Pull request merged**